[Brian Chow] iP

brian16600 opened the pull request:
# Duke

Duke is a interactive chatbot with UI. 
> "Your mind is for having ideas, not holding them" - David Allen

Duke is a task tracker that can track three categories of tasks:
- Todo
- Deadline
- Event

## Setting up

Setting up Duke is extremely easy. All you need to do is:
1) Download JDK 11 and above from [here](https://www.jetbrains.com/help/idea/sdk.html#set-up-jdk)
2) Download the latest JAR file release from the right side of this repository
3) Run the command java -jar ip.jar

It's that simple! :)

## Commands
There are a few commands you can run in Duke. These include:
- [x] todo
- [X] deadline
- [X] event
- [X] delete
- [X] list
- [X] find
- [X] bye

## Using an IDE
You can further edit/modify Duke by navigating to the `src` folder which contains all the `java` files

You can run it from Launcher.java by using __main__:
<pre>
public class Launcher {
    <b>public static void main(String[] args)</b> {
        Application.launch(Main.class,args);
    }
}
</pre>

Commit:
Add Gradle support

brian16600 committed:
Level-1

brian16600 committed:
Level-2. Changed the return values of levelOneRespond and implemented levelTwoRespond

brian16600 committed:
Level-3

brian16600 committed:
Level-3

brian16600 committed:
Level-3

brian16600 committed:
Level-4.
Added Event, Deadline and Todo subclasses of Task. Updated if-else in levelFourRespond to handle the 3 subclasses.

brian16600 committed:
Level-5.
Created DukeException class to handle incorrect input length/commands. Shifted if-else loop to a switch-case in InputHandler.java,

brian16600 committed:
Level-6
Added recommended inputs for exceptions and added delete case in InputHandler.java.

brian16600 committed:
Level-7
Rewrote code into InputHandler.java. Added Storage.java to handle the storage of tasks as well as writing to data.txt file for saving.

declanleeler commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,60 @@
+ * Represents a Task. Contains a Task constructor, two methods to mark and unmark tasks, toString() method as well as a isMark() method to check if Task is marked
+ */
+public class Task {
+    private boolean mark;
```
Not sure I like the naming of the boolean variable. As much as possible, use a prefix such as is, has, was, etc. for boolean variable/method names so that linters can automatically verify that this style rule is being followed. Perhaps it would be better to name it as follows?
```suggestion
    private boolean isMarked;
```

declanleeler commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,60 @@
+    /**
+     * markTask as done
+     */
+    public void markTask () {
```
Should this be named in this manner? Perhaps the following would comply with the coding standards better.
```suggestion
    public void setMarkedTask () {
```

declanleeler commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,60 @@
+    /**
+     * unmarkTask
+     */
+    public void unmarkTask() {
```
Should this be named in this manner? Perhaps the following would comply with the coding standards better.
```suggestion
    public void setUnmarkedTask () {
```

declanleeler commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,60 @@
+     *
+     * @return boolean on whether task is marked
+     */
+    public boolean isMark() {
```
Should this be changed to another name, if you do decide to accept the suggestion of the boolean variable naming?  Perhaps the following would work.
```suggestion
    public void hasBeenMarked () {
```

declanleeler commented on src/main/java/InputHandler.java:
```diff
@@ -0,0 +1,127 @@
+        String[] splitInput = input.split(" ");
+        String inputCommand = splitInput[0];
+        switch (inputCommand) {
+            case "todo":
```
I didn't like the indentation after the switch, are you sure this complies with the coding standard? Consider leaving no indentation for the case statements.

declanleeler reviewed (commented):
I found the code easy to read and follow. There are a couple of coding standard violations. The InputHandler class is also very long and could be shortened with the use of more classes to split up the code in the switch-case and if-else.

brian16600 committed:
Level-8
Changed multiple variable names to better suit their purposes. Added ability to take in specific date/time for Event/Deadline classes in yyyy-mm-dd hh:mm format and convert it. Added handling for DateTimeParseException

xMashedxTomatox commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,20 @@
+    private String dueDate;
+    public Deadline(String name, String time) {super(name); this.dueDate = time;}
+
+    /**
+     * @override
+     * @return String of Deadline task, eg [D][X] Deadline (by:XX)
+     */
```
@Override should not be in the javadoc. Missing description for the method in the javadoc.
Recommendation
```
/**
* Prints the String representation for the Deadline task.
* 
* @return String of Deadline task, eg [D][X] Deadline (by:XX)
* /
@Override
public String toString() {...}
```

xMashedxTomatox replied:
Override should not be in the javadoc. Missing method's description for the javadoc.
Recommendation:
```
/**
* Prints the String representation for the Deadline task.
* 
* @return String of Deadline task, eg [D][X] Deadline (by:XX)
* /
@Override
public String toString() {...}
```

xMashedxTomatox commented on src/main/java/DukeException.java:
```diff
@@ -0,0 +1,12 @@
+public class DukeException extends Exception{
```
Missing javadoc for class.

xMashedxTomatox replied:
Missing javadoc for class.

xMashedxTomatox commented on src/main/java/DukeException.java:
```diff
@@ -0,0 +1,12 @@
+public class DukeException extends Exception{
+    private String errorMessage;
+    public DukeException (String errorMessage) {
```
Missing javadoc for constructor.

xMashedxTomatox commented on src/main/java/DukeException.java:
```diff
@@ -0,0 +1,12 @@
+        this.errorMessage = errorMessage;
+    }
+
+    /**
+     * @override
+     * @return String errorMessage
+     */
```
Same as above.

xMashedxTomatox commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,18 @@
+    private String dueDate;
+    public Event (String name, String time) { super(name); this.dueDate = time;}
+
+    /**
+     * @override
+     * @return String of Event task, eg: [E][X] Event
+     */
```
Same as above.

xMashedxTomatox commented on src/main/java/InputHandler.java:
```diff
@@ -0,0 +1,127 @@
+ * Processes the input into 7 categories: Todo, Event, Deadline, list, mark, unmark, bye and throws error
+ */
+public class InputHandler {
+    private ArrayList<Task> arr;
```
Might be better to have a more specific and plural name for a list of task.
Recommendation:
`
private ArrayList<Task> tasks;
`

xMashedxTomatox replied:
Might be better to have a more specific and plural name for a list of task.
Recommendation:
```
private ArrayList<Task> tasks;
```

xMashedxTomatox commented on src/main/java/InputHandler.java:
```diff
@@ -0,0 +1,127 @@
+
+
+    /**
+     *
```
Missing method description for javadoc.

xMashedxTomatox replied:
Missing method description for javadoc.

xMashedxTomatox commented on src/main/java/InputHandler.java:
```diff
@@ -0,0 +1,127 @@
+                    throw new DukeException("Wrong usage of list! Correct usage: list");
+                }
+            case "mark":
+                if (splitInput.length == 2) {
+                    System.out.println("Nice! I've marked this task as done:\n");
+                    int idx = Integer.parseInt(splitInput[1]) - 1;
+                    Task taskToBeMarked = arr.get(idx);
+                    taskToBeMarked.markTask();
+                    return false;
+                } else {
+                    throw new DukeException("Wrong usage of mark! Correct usage: mark [index]");
+                }
+            case "unmark":
+                if (splitInput.length == 2) {
+                    System.out.println("OK, I've marked this task as not done yet:\n");
+                    int idx = Integer.parseInt(splitInput[1]) - 1;
+                    Task taskToBeUnmarked = arr.get(idx);
+                    taskToBeUnmarked.unmarkTask();
+                    return false;
+                } else {
+                    throw new DukeException("Wrong usage of unmark! Correct usage: unmark [index]");
+                }
+            case "delete":
+                if (splitInput.length == 2) {
+                    int idx = Integer.parseInt(splitInput[1]) - 1;
+                    Task taskToBeDeleted = arr.get(idx);
+                    arr.remove(idx);
+                    System.out.println("Noted. I've removed this task:\n" + taskToBeDeleted + "\nNow you have " + arr.size() + " tasks in the list");
+                    return false;
+                } else {
+                    throw new DukeException("Wrong usage of delete! Correct usage: delete [index]");
+                }
+            case "bye":
+                return true;
+            default:
+                throw new DukeException(":( OOPS!!! I'm sorry, but I don't know what that means! Possible commands: todo [task], event [task] /at [time],"
+                + " deadline [task] /by [time], mark [index], unmark [index], delete [index], bye");
+        }
+
+    }
```
This method severely exceeds the recommended number of lines in a method (30). 
Recommendation:
Can try to abstract out the details of each case statement into a different method.

xMashedxTomatox commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,60 @@
+    public String name;
+
+    /**
+     * Constructor
```
Might be better to have a more specific description of a constructor instead of simply "Constructor".
Recommendation:
Creates a task object with the given parameter as the name of the task.

xMashedxTomatox replied:
Might be better to have a more specific description of a constructor instead of simply "Constructor".
Recommendation:
```
/**
* Creates a task object with the given parameter as the name of the task.
* @param name name of the task
*/

```

xMashedxTomatox reviewed (approved):
Looks pretty neat overall, LGTM! Just need to clean up a little on the javadocs and abstracting out certain logic to make the code a little bit better to read.

xMashedxTomatox commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,20 @@
+ * Includes a dueDate attribute. Overrides toString() from Task
+ */
+public class Deadline extends Task {
+    private String dueDate;
+    public Deadline(String name, String time) {super(name); this.dueDate = time;}
```
Missing javadocs for constructor.

Dineshraj555 replied:
Could leave spacing for better readability.

xMashedxTomatox commented on src/main/java/InputHandler.java:
```diff
@@ -0,0 +1,127 @@
+                    return false;
+                } else {
+                    throw new DukeException("Wrong usage of list! Correct usage: list");
+                }
+            case "mark":
+                if (splitInput.length == 2) {
+                    System.out.println("Nice! I've marked this task as done:\n");
+                    int idx = Integer.parseInt(splitInput[1]) - 1;
+                    Task taskToBeMarked = arr.get(idx);
+                    taskToBeMarked.markTask();
+                    return false;
+                } else {
+                    throw new DukeException("Wrong usage of mark! Correct usage: mark [index]");
+                }
+            case "unmark":
+                if (splitInput.length == 2) {
+                    System.out.println("OK, I've marked this task as not done yet:\n");
+                    int idx = Integer.parseInt(splitInput[1]) - 1;
+                    Task taskToBeUnmarked = arr.get(idx);
+                    taskToBeUnmarked.unmarkTask();
+                    return false;
+                } else {
+                    throw new DukeException("Wrong usage of unmark! Correct usage: unmark [index]");
+                }
+            case "delete":
+                if (splitInput.length == 2) {
+                    int idx = Integer.parseInt(splitInput[1]) - 1;
+                    Task taskToBeDeleted = arr.get(idx);
+                    arr.remove(idx);
+                    System.out.println("Noted. I've removed this task:\n" + taskToBeDeleted + "\nNow you have " + arr.size() + " tasks in the list");
+                    return false;
+                } else {
+                    throw new DukeException("Wrong usage of delete! Correct usage: delete [index]");
+                }
+            case "bye":
+                return true;
+            default:
+                throw new DukeException(":( OOPS!!! I'm sorry, but I don't know what that means! Possible commands: todo [task], event [task] /at [time],"
+                + " deadline [task] /by [time], mark [index], unmark [index], delete [index], bye");
+        }
```
This method severely exceeds the recommended number of lines in a method (30).
Recommendation:
Can try to abstract out the details of each case statement into a different method.

xMashedxTomatox reviewed (commented):
Looks pretty neat overall, LGTM! Just need to clean up a little on the javadocs and abstracting out certain logic to make the code a little bit better to read.

Dineshraj555 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,40 @@
+
+
 }
+
+
+
+
+
+
+
+
+
```
Could clear by these empty spaces.

ianfromdover replied:
Agreed

Dineshraj555 commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,18 @@
+ */
+public class Event extends Task {
+    private String dueDate;
+    public Event (String name, String time) { super(name); this.dueDate = time;}
```
This do not follow the module coding standards.

Dineshraj555 commented on src/main/java/InputHandler.java:
```diff
@@ -0,0 +1,127 @@
+            case "bye":
+                return true;
+            default:
+                throw new DukeException(":( OOPS!!! I'm sorry, but I don't know what that means! Possible commands: todo [task], event [task] /at [time],"
```
Could add indentation.

Dineshraj555 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.Arrays;
+
+/**
+ * Represents a Task. Contains a Task constructor, two methods to mark and unmark tasks, toString() method as well as a isMark() method to check if Task is marked
```
Could add indentation.

Dineshraj555 reviewed (commented):
Overall, code was easy to read. Just minor coding standard violations to take down note of. Good job.

brian16600 committed:
Added TaskList and Parser classes. TaskList contains the arraylist of tasks with methods and Parser makes sense of user input for Duke

brian16600 committed:
Added Parser, Storage and Tasklist

brian16600 committed:
Added more OOP and documentation

brian16600 committed:
Added packages: duke.duke, duke.ui, duke.storage and duke.task

brian16600 committed:
Merge branch 'add-gradle-support'

brian16600 committed:
Added add-gradle-support branch by merging into master

brian16600 committed:
Ensured javadoc documentation is complete for all methods

brian16600 committed:
Added InputHandlerTest.java and TaskListTest.java

brian16600 committed:
Added NumberFormatException handling for delete, unmark and mark if not in correct format. Added proper javadocs documentation

brian16600 committed:
Changed formatting for coding standards based on feedback

brian16600 committed:
Added find and NumberFormatException in cases where non-integer is passed in.

ianfromdover commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,36 @@
+        this.dueTime = LocalTime.parse(time);
+    }
+
+    /**
+     * @override
+     * @return String of Deadline task, eg [D][X] Deadline (by:XX) vs [D][✓] Deadline (by;XX)
+     */
+
```
Perhaps you might want to place the JavaDoc directly above the method without a spacing, as well as move the override outside for it to be in effect :)

ianfromdover commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,35 @@
+public class Event extends Task {
+    public LocalDate dueDate;
+    public LocalTime dueTime;
+    public Event (String name, String date) throws DateTimeParseException {
+        super(name);
+        this.dueDate = LocalDate.parse(date);
+        this.dueTime = null;
+    }
+    public Event (String name, String date, String time) throws DateTimeParseException {
+        super(name);
+        this.dueDate = LocalDate.parse(date);
+        this.dueTime = LocalTime.parse(time);
+    }
```
Might want to include a line of space between the attributes and between methods :)

ianfromdover commented on src/main/java/InputHandler.java:
```diff
@@ -0,0 +1,118 @@
+                this.storage.writeData(newDeadline);
+                printAddTaskMessage(newDeadline);
+                return false;
+            } else {
+                throw new DukeException(":( OOPS!!! The description of a deadline cannot be empty. Correct usage: deadline [task] /by [time]");
+            }
+        case "list":
+            if (splitInput.length == 1) {
+                parser.parse(CommandType.LIST, this.storage, splitInput);
+                return false;
+            } else {
+                throw new DukeException("Wrong usage of list! Correct usage: list");
+            }
+        case "mark":
+            if (splitInput.length == 2) {
+                parser.parse(CommandType.MARK, this.storage, splitInput);
+                return false;
+            } else {
+                throw new DukeException("Wrong usage of mark! Correct usage: mark [index]");
+            }
+        case "unmark":
+            if (splitInput.length == 2) {
+                parser.parse(CommandType.UNMARK, this.storage, splitInput);
+                return false;
+            } else {
+                throw new DukeException("Wrong usage of unmark! Correct usage: unmark [index]");
+            }
+        case "delete":
+            if (splitInput.length == 2) {
+                parser.parse(CommandType.DELETE, this.storage, splitInput);
+                return false;
+            } else {
+                throw new DukeException("Wrong usage of delete! Correct usage: delete [index]");
+            }
+        case "bye":
+            return true;
+        default:
+            throw new DukeException(":( OOPS!!! I'm sorry, but I don't know what that means! Possible commands: todo [task], event [task] /at [time],"
+            + " deadline [task] /by [time], mark [index], unmark [index], delete [index], bye");
+        }
```
Good use of switch statement to clean up code. Easy to follow 👍

ianfromdover commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,84 @@
+
+public class Parser {
+
+    public Task parse(InputHandler.CommandType type, String[] splitInput) throws DukeException {
```
Perhaps it may be easier to understand if the 2 parse functions were combined into 1 :)

ianfromdover commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,146 @@
+    public String taskToStringConverter(Task task) {
+        String output = "";
+        if (task instanceof Todo) {
+            String mark = (task.hasBeenMarked()) ? "[✓]" : "[X]";
+            output = "[T] " + mark + " / " + task.name + "\n";
+        } else if (task instanceof Deadline deadline) {
+            String mark = (deadline.hasBeenMarked()) ? "[✓]" : "[X]";
+            output = "[D] " + mark + " / " + deadline.name + " / " + deadline.dueDate + " / " + deadline.dueTime + "\n";
+        } else if (task instanceof Event event) {
+            String mark = (event.hasBeenMarked()) ? "[✓]" : "[X]";
```
The ticks are a nice touch 😊👌

ianfromdover commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,146 @@
+        String listOfTasks = "";
+        for (int i = 1; i <= this.taskList.size(); i++) {
+            Task task = this.taskList.get(i - 1);
+            if (task.hasBeenMarked()) {
+                listOfTasks += i + ". " + task + "\n";
+            } else {
+                listOfTasks += i + ". " + task + "\n";
+            }
```
Good use of spaces between operators 👍

ianfromdover reviewed (commented):
Neat and organised codebase, variables are well named. Good job 👍

brian16600 committed:
Added JAR functionality

brian16600 committed:
Shifted into packages

brian16600 committed:
revert

brian16600 committed:
Tentative changes

brian16600 committed:
Adding Duke GUI

brian16600 committed:
Add Assertions

Add Assertions for index out of bounds exception for delete/mark/unmark commands.

Had to handle this exception for user error, change made addresses this.

It is done using assertions to ensure index is within bounds.

Change some of the code behind the bye command and storage.

GUI was not exiting properly when bye was called, and some of the writing to storage had issues.

Add a new method rewriteData() to address this issue and fixed bye command.

Change storage writing for marked vs unmarked to 1(marked) vs 0(unmarked), since the GUI had issues displaying the tick.

brian16600 committed:
Clean up text-ui-test

brian16600 committed:
Tentative changes to code quality

brian16600 committed:
Improve Code Quality based on Issue Tracker

Update and add documentation for all methods.

Improve abstraction throughout the code by abstracting out many Strings for return statements. Some methods reused the same String, hence abstracting them out makes the code cleaner.

Shorten the method length for methods in Parser.java, InputHandler.java and Storage.java as method lengths were too long and add new methods for each case in the switch to handle the code. This was aimed at making the code cleaner.

brian16600 committed:
Merge pull request #2 from brian16600/branch-A-CodeQuality

Code Quality

brian16600 committed:
Update README.md

brian16600 committed:
Update README.md

brian16600 committed:
Create README.md

brian16600 committed:
Update README.md

brian16600 committed:
Update README.md

brian16600 committed:
Update README.md

brian16600 committed:
Update README.md

brian16600 committed:
Update README.md

brian16600 committed:
Update README.md

brian16600 committed:
Temporary addition of Snooze.

brian16600 committed:
Add snooze command

Add snooze command for the type B Extension. This functionality was added as an 8th command type.

brian16600 committed:
Merge branch 'master' of https://github.com/brian16600/ip

brian16600 committed:
Add Ui.png

brian16600 committed:
Add Ui.png

brian16600 committed:
Update README.md

brian16600 committed:
Update README.md

brian16600 committed:
Update README.md

brian16600 committed:
Update README.md

brian16600 committed:
Update README.md
